api/src/main/java/io/opentelemetry/common/Attributes.java

```diff
@@ -0,0 +1,358 @@
+    public AttributeValue getValue(String key) {
+      List<Object> data = data();
+      for (int i = 0; i < data.size(); i++) {
+        if (data.get(i++).equals(key)) {
+          return (AttributeValue) data.get(i);
+        }
+      }
+      return null;
+    }
+
+    @Override
+    public Iterator<Entry<String, AttributeValue>> iterator() {
+      return new ArrayIterator();
+    }
+
+    private class ArrayIterator implements Iterator<Entry<String, AttributeValue>> {
+      private int currentIndex = 0;
+
+      @Override
+      public boolean hasNext() {
+        return data().size() > currentIndex;
+      }
+
+      @Override
+      public Entry<String, AttributeValue> next() {
+        if (!hasNext()) {
+          throw new IllegalStateException("no more");
+        }
+        List<Object> data = data();
+        return new SimpleImmutableEntry<>(
+            (String) data.get(currentIndex++), (AttributeValue) data.get(currentIndex++));
+      }
+
+      @Override
+      public void remove() {
+        throw new UnsupportedOperationException("remove");
+      }
+    }
+  }
+
+  /** javadoc me. */
+  public static class Builder {
+    private final List<Object> data = new ArrayList<>();
+
+    /** javadoc me. */
+    public Attributes build() {
+      return sortAndFilter(Arrays.asList(data.toArray()));
+    }
+
+    /** javadoc me. */
+    public Builder addAttribute(String key, String value) {
+      data.add(key);
+      data.add(AttributeValue.stringAttributeValue(value));
+      return this;
+    }
+
+    /** javadoc me. */
+    public Builder addAttribute(String key, long value) {
+      data.add(key);
+      data.add(AttributeValue.longAttributeValue(value));
+      return this;
+    }
+
+    /** javadoc me. */
+    public Builder addAttribute(String key, double value) {
+      data.add(key);
+      data.add(AttributeValue.doubleAttributeValue(value));
+      return this;
+    }
+
+    /** javadoc me. */
+    public Builder addAttribute(String key, boolean value) {
+      data.add(key);
+      data.add(AttributeValue.booleanAttributeValue(value));
+      return this;
+    }
+
+    /** javadoc me. */
+    public Builder addAttribute(String key, String... value) {
+      data.add(key);
+      data.add(AttributeValue.arrayAttributeValue(value));
+      return this;
+    }
+
+    /** javadoc me. */
+    public Builder addAttribute(String key, Long... value) {
+      data.add(key);
+      data.add(AttributeValue.arrayAttributeValue(value));
+      return this;
+    }
+
+    /** javadoc me. */
+    public Builder addAttribute(String key, Double... value) {
+      data.add(key);
+      data.add(AttributeValue.arrayAttributeValue(value));
+      return this;
+    }
+
+    /** javadoc me. */
+    public Builder addAttribute(String key, Boolean... value) {
+      data.add(key);
+      data.add(AttributeValue.arrayAttributeValue(value));
+      return this;
+    }
+  }
+
+  private static class EmptyAttributes extends Attributes {
```

Comment:
Would use the enum pattern :)

Reply:
Hm. In this case, since I made Attributes an abstract class (rather than an interface, so I could add static methods to it for java7), I can't use an enum, because they can't extend classes.